Add support for exploding dice

LiruMouse opened the pull request:
Exploding dice reroll the current die when it rolls as high as it can, see the notes in the second commit for more.
The syntax works by punctuating a dice expression with an exclamation mark.
This won't work for d1 because of the obvious loop and this won't work with keeping because the features don't make sense together.
I wasn't able to figure out how to mock multiple roll results so I haven't implemented tests yet that have actual explosions.

LiruMouse committed:
Remove unused variable

LiruMouse committed:
Add support for Exploding dice

Exploding dice roll again
They're denoted with ! (See https://en.wikipedia.org/wiki/Dice_notation)

Also fixes tests to work with the new syntax
Doubling up exclamation marks allows one to be parsed and the other not.

It doesn't add tests because I wasn't sure quite where
and how to mock sequential rolls

cidthecoatrack commented on RollGen/Expressions/AlbatrossExpressionEvaluator.cs:
```diff
@@ -6,8 +6,6 @@ namespace RollGen.Expressions
 {
     internal class AlbatrossExpressionEvaluator : ExpressionEvaluator
     {
-        private const string CommonRollRegexPattern = "d *\\d+(?: *k *\\d+)?";
```
Nice catch

cidthecoatrack commented:
Hey Lirusaito,

Sorry for the SUPER DELAYED response.  Work has kinda eaten me alive, and I've been deep in the weeds for CreatureGen.  I like this PR in general, just have a few questions/requests:

1. To make sure that I understand the "explode" functionality correctly: if I roll `1d6!`, and I roll 1, 2, 3, 4, or 5, then my total is what I rolled.  HOWEVER, if I roll a 6, then it effectively becomes `1d6+6`, and will do that until I don't roll a 6 (`1d6+12`, `1d6+18`, etc.).  Is that correct?
2. I will definitely need the unit tests before I can approve the PR.  It's been a while since I have been in the codebase for this project, therefore I have also forgotten how to mock multiple rolls.  However, I am sure there are unit tests that do this, so if you look through the test projects, I am sure you can see an example there.
3. Are `keep` and `explode` mutually exclusive?  I feel like `keep` just becomes a hard limit on the explosion.  For example, if I roll `6d6k4!`, and I rolled `[4, 6, 5, 4, 3, 6, 6, 1, 1]` (with `explode` giving us the extra rolls), then when `keep` performs, it will take `[6, 6, 6, 5]` for a total of `23`.  Of course, this assumes that I understand the explode functionality correctly.

Unit tests I will want to see:
  * Keep/Explode exceptions thrown when intermingling (or new functionality, depending)
  * Test where you try to explode `1dX`, roll number less than `X`
  * Test where you try to explode `1dX`, roll `X` once
  * Test where you try to explode `1dX`, roll `X` twice (but total should be `X`)
  * Test where you explode `YdX`, roll number less than `X`
  * Test where you explode `YdX`, roll X once
  * Test where you explode `YdX`, roll X twice in a row
  * Test where you explode `YdX`, roll X twice, not in a row
  * A "plaintext" version of explode (effectively a test that the regex pattern finds what it should, similar to the one you replaced where the sentence with `!` now ends with `.`)
  *

cidthecoatrack commented:
Also, could you merge into `develop` instead of `master`?  That way once it's in, I can manage the actual release/NuGet deployment

Lirusaito commented:
> Sorry for the SUPER DELAYED response. Work has kinda eaten me alive, and I've been deep in the weeds for CreatureGen. I like this PR in general, just have a few questions/requests:
That's totally fine, I assumed you were otherwise occupied. I've had a look at the site recently and it's coming along quite nicely.

>     1. To make sure that I understand the "explode" functionality correctly: if I roll `1d6!`, and I roll 1, 2, 3, 4, or 5, then my total is what I rolled.  HOWEVER, if I roll a 6, then it effectively becomes `1d6+6`, and will do that until I don't roll a 6 (`1d6+12`, `1d6+18`, etc.).  Is that correct?
Yes

>     2. I will definitely need the unit tests before I can approve the PR.  It's been a while since I have been in the codebase for this project, therefore I have also forgotten how to mock multiple rolls.  However, I am sure there are unit tests that do this, so if you look through the test projects, I am sure you can see an example there.
I'll start looking

>     3. Are `keep` and `explode` mutually exclusive?  I feel like `keep` just becomes a hard limit on the explosion.  For example, if I roll `6d6k4!`, and I rolled `[4, 6, 5, 4, 3, 6, 6, 1, 1]` (with `explode` giving us the extra rolls), then when `keep` performs, it will take `[6, 6, 6, 5]` for a total of `23`.  Of course, this assumes that I understand the explode functionality correctly.
They are mutually exclusive, in an expression only the first one will be found and evaluated, and in the fluid syntax, it will throw. I did this because I cannot find an example of both being used in the wild and I wouldn't know whether to keep the highest rolls only or keep the highest dice and all of their explosions too.
 
> Unit tests I will want to see:
I'll get to work on those, should hopefully have them done today, if not, by the end of the week.

Lirusaito commented:
Oh, real quick, what did you mean by this?
>Test where you try to explode 1dX, roll X twice (but total should be X)

Total should be X? wouldn't the total be 2X+Y where Y is between 1 and X-1?

LiruMouse committed:
Remove unused member variable

LiruMouse committed:
Fix this test another way, keeping the exclamation mark

LiruMouse committed:
[Explode] Added requested unit tests

Test for intermingling Keep/Explode throwing exceptions
Test for not exploding endlessly when rolling a d1
Test where you try to explode 1dX, roll number less than X
Test where you try to explode 1dX, roll X once
Test where you try to explode 1dX, roll X twice
Test where you explode YdX, roll number less than X
Test where you explode YdX, roll X once
Test where you explode YdX, roll X twice in a row
Test where you explode YdX, roll X twice, not in a row
Tests for plaintext recognition of explode in regex

cidthecoatrack commented on RollGen.Tests.Integration.IoC/ExplodeTests.cs:
```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Collections.Generic;
+
+namespace RollGen.Tests.Unit
```
Why is this test class in the integration test project?

LiruMouse replied:
I think I needed to link against libraries already referenced there. Also I wasn't quite sure where this test should be put, anyhow.

cidthecoatrack replied:
If you can move this into the unit test project, then I can merge this pull request

LiruMouse replied:
Sure give me a bit

cidthecoatrack commented:
Ah, thank you.  I actually meant test that if I try to roll `Xd1`, that it doesn't explode.  Thanks for catching that!

LiruMouse committed:
Move ExplodeTests into the unit test project

cidthecoatrack commented on RollGen.Tests.Unit/ExplodeTests.cs:
```diff
@@ -33,7 +35,7 @@ public class ExplodeTests : IntegrationTests
         {
             var random = new MockRandom(rolls);
-            var partialRoll = new NumericPartialRoll(quantity, random, GetNewInstanceOf<ExpressionEvaluator>());
+            var partialRoll = new NumericPartialRoll(quantity, random, evaluator);
```
Is there a reason you made your `MockRandom` class, instead of using `Mock<Random>`?  Also, you should be using a mock of the `ExpressionEvaluator` as well

LiruMouse replied:
I wanted to be able to provide an array to return in order, as far as I could tell `Mock<Random>` would only return A single value, did I miss something you're aware of?

As for the expression evaluator, we want to be using albatross, we're not mocking it because we want it to work as it would normally.

cidthecoatrack replied:
With `Moq`, you can set up a return sequence that will emulate your array.  Something like

```
mockRandom.SetupSequence(r => r.Next())
    .Returns(1)
    .Returns(0)
    .Returns(9000);
```

As for the expression evaluator, in reality we're not testing the albatross engine here, just our own explode logic.  However, setting up the mock expression evaluator would be somewhat involved, so I guess it would be easier to just use the real thing in this case.  If I feel the need, I'll refactor it later.

LiruMouse replied:
Awesome, I pushed a fix to use a mocked return sequence.

Yeah, that was the way I felt about the evaluator, ideally we could mock it but we need good responses... I suppose we could, for each- Wait, we don't even use the evaluator for this, because we build the expression and we can parse it on our own, I can just mock it!
I'll push that in a bit.

LiruMouse replied:
There we go. And I was thinking we could pass into the tests, through test parameters, what it should be after evaluation, should we need the evaluator to be mocked for real.

LiruMouse committed:
Use Moq<Random> with a sequence instead of custom MockRandom class

LiruMouse committed:
Mock unused expression evaluator instead of using Albatross

LiruMouse committed:
Merge branch 'develop' into Explode

cidthecoatrack commented:
Awesome.  I'll merge this, do version/deployment prep, and then release this as an update to NuGet. :)
